archive/imo/imo1979_q1.lean

```diff
@@ -0,0 +1,302 @@
+/-
+Copyright (c) 2020 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+
+import tactic
+import algebra.big_operators.ring
+import data.nat.parity
+import data.finset.intervals
+import data.padics.padic_norm
+
+/-!
+# IMO 1979 Q1
+
+Let `p` and `q` be positive integers such that
+
+`p/q=1-1/2+1/3-1/4+...-1/1318+1/1319`
+
+Prove that `p` is a multiple of 1979.
+
+## The solution
+
+The proof we formalise is the following. Rewrite the sum as
+`1+1/2+1/3+...+1/1319 - 2*(1/2+1/4+1/6+...+1/1318)=1/660+1/661+...+1/1319
+and now re-arranging as (1/660+1/1319)+(1/661+1/1318)+... we see that
+the numerator of each fraction is 1979 and the denominator is coprime to 1979
+(note that 1979 is prime). Hence the 1979-adic valuation of each fraction is positive,
+and thus the 1979-adic valuation of the sum is also positive.
+
+## Remarks on the formalisation
+
+The p-adic valuation function on ℚ is ℤ-valued and hence has v(0)=0 (rather than +∞);
+we thus had to occasionally deal with funny edge cases which aren't there mathematically.
+In retrospect it might have been easier to work with p-adic norms, which are ℚ-valued
+and don't have this problem.
+-/
+
+
+open_locale big_operators
+
+open finset
+
+open nat
+
+instance : fact (nat.prime 1979) := by norm_num
+
+open finset
+
+namespace imo1979q1
+
+-- some constants
+@[reducible] def a : ℚ := ∑ n in range 1320, (-1)^(n + 1) / n
+@[reducible] def b : ℚ := ∑ n in range 1320, 1 / n
+@[reducible] def c : ℚ := ∑ n in range 660, 1 / n
+@[reducible] def d : ℚ := ∑ n in Ico 660 1320, 1 / n
+@[reducible] def e : ℚ := ∑ n in range 330, 1 / (n + 660)
+@[reducible] def f : ℚ := ∑ n in range 330, 1 / (1319 - n)
+
+/-
+  The goal is equivalent to showing that the 1979-adic valuation of a is positive.
+
+  We start with some lemmas:
+
+  Lemma 1 : b - a = c
+  Lemma 2 : c + d = b
+  Corollary 3 : a = d
+  Lemma 4 : d = e + f
+
+  Then we prove the theorem, by showing the 1979-adic
+  valuation of e + f is positive.
+-/
+
+@[reducible] def double : ℕ ↪ ℕ := ⟨λ n, 2 * n, λ a b, mul_left_cancel' (by norm_num)⟩
+
+lemma lemma1 : b - a = c :=
+  calc b - a = ∑ n in range 1320, (1/n - (-1)^(n+1)/n) : by rw sum_sub_distrib
+    ... = ∑ n in range 1320, ite (even n) (2/n) 0 : by {
+      apply sum_congr rfl,
+      rintros x hx,
+      rw [pow_add, pow_one, mul_neg_one, neg_div, sub_neg_eq_add, ← _root_.add_div],
+      split_ifs,
+      { rcases h with ⟨n, rfl⟩,
+        rw [pow_mul, pow_two],
+        simp, norm_num },
+      { rw not_even_iff at h,
+        have h8 := nat.mod_add_div x 2,
+        rw h at h8,
+        suffices : (1 : ℚ) + (-1)^x = 0,
+          rw this, simp,
+        rw ← h8,
+        simp [pow_add, pow_mul],
+      }
+    }
+    ... = ∑ (x : ℕ) in filter even (range 1320), 2 / x : by rw sum_filter
+    ... = ∑ (x : ℕ) in map double (range 660), (2 : ℚ) / x : by {
```

Comment:
Why is `(2 : ℚ)` needed here but not on the previous line?

Reply:
Thanks. It's not,